### ISSUE TITLE
ORA-06550 when calling a PL/SQL procedure with a %ROWTYPE parameter that has DATE fields when dateAsTimestamp=true

### ISSUE THREAD

**Issue by mnisius** [jOOQ/jOOQ#12019]
Hi, 

I know Oracle's DATE data type does not conform to the SQL standard however we have a large legacy Database and each Table has DATE columns and it's a requirement to insert a time as well in this fields. 

This is the reason why we use dateAsTimestamp=true in all our java Projects. This works great for SQL but each PLSQL Procedure with a ROWTYPE that has a Date Column will throw an error. The procedures are working fine if I set dateAsTimestamp to false but then I cannot use them because I can't insert a time information. 

### Expected behavior

Procedure is called with no error

### Actual behavior
Error Message:

```
org.jooq.exception.DataAccessException: SQL [declare type rowtype1 is record("ASMV_ID" number(19), "ASMV_SENDER" varchar2(1000), "ASMV_EMPFAENGER" varchar2(4000), "ASMV_AUSFUEHRUNGSZEITPUNKT" timestamp, "ASMV_CC" varchar2(4000), "ASMV_BCC" varchar2(4000), "ASMV_HEADER" varchar2(1000), "ASMV_TEXT" clob, "ASMV_ERSTELLT_VON" varchar2(20), "ASMV_ERSTELLUNGSZEITPUNKT" timestamp, "ASMV_MIME_TYPE" varchar2(200), "ASMV_PFAD" varchar2(100), "ASMV_DATEINAME" varchar2(100), "ASMV_BLOB" blob, "ASMV_EXT_ANBIETER" varchar2(1000), "ASMV_STATUS" varchar2(1), "ASMV_SMT_ID" number(19)); type rowtype2 is record("RUP_ID" number(19), "RUP_EREIGNISTYP" varchar2(1), "RUP_EREIGNISCODE" varchar2(30), "RUP_EREIGNISTEXT" varchar2(4000), "RUP_ERSTELLT_VON" varchar2(20), "RUP_ERSTELLUNGSZEITPUNKT" timestamp, "RUP_USERNAME" varchar2(20), "RUP_APPLIKATION" varchar2(200), "RUP_PROGRAMMNAME" varchar2(200), "RUP_PROGRAMMLAUF_ID" number(19), "RUP_PROZEDURNAME" varchar2(200), "RUP_CODEBLOCK" varchar2(4000), "RUP_ENTITAET" varchar2(200), "RUP_PARAMETER" varchar2(4000), "RUP_PROZESS" varchar2(200), "RUP_TEILPROZESS" varchar2(200), "RUP_SCHLUESSELTYP" varchar2(200), "RUP_SCHLUESSEL" varchar2(4000), "RUP_PROGRAMMSTACK" varchar2(4000), "RUP_BRACKET" varchar2(200)); "r1" rowtype1; "r2" rowtype2; begin "r1"."ASMV_ID" := ?; "r1"."ASMV_SENDER" := ?; "r1"."ASMV_EMPFAENGER" := ?; "r1"."ASMV_AUSFUEHRUNGSZEITPUNKT" := ?; "r1"."ASMV_CC" := ?; "r1"."ASMV_BCC" := ?; "r1"."ASMV_HEADER" := ?; "r1"."ASMV_TEXT" := ?; "r1"."ASMV_ERSTELLT_VON" := ?; "r1"."ASMV_ERSTELLUNGSZEITPUNKT" := ?; "r1"."ASMV_MIME_TYPE" := ?; "r1"."ASMV_PFAD" := ?; "r1"."ASMV_DATEINAME" := ?; "r1"."ASMV_BLOB" := ?; "r1"."ASMV_EXT_ANBIETER" := ?; "r1"."ASMV_STATUS" := ?; "r1"."ASMV_SMT_ID" := ?; "r2"."RUP_ID" := ?; "r2"."RUP_EREIGNISTYP" := ?; "r2"."RUP_EREIGNISCODE" := ?; "r2"."RUP_EREIGNISTEXT" := ?; "r2"."RUP_ERSTELLT_VON" := ?; "r2"."RUP_ERSTELLUNGSZEITPUNKT" := ?; "r2"."RUP_USERNAME" := ?; "r2"."RUP_APPLIKATION" := ?; "r2"."RUP_PROGRAMMNAME" := ?; "r2"."RUP_PROGRAMMLAUF_ID" := ?; "r2"."RUP_PROZEDURNAME" := ?; "r2"."RUP_CODEBLOCK" := ?; "r2"."RUP_ENTITAET" := ?; "r2"."RUP_PARAMETER" := ?; "r2"."RUP_PROZESS" := ?; "r2"."RUP_TEILPROZESS" := ?; "r2"."RUP_SCHLUESSELTYP" := ?; "r2"."RUP_SCHLUESSEL" := ?; "r2"."RUP_PROGRAMMSTACK" := ?; "r2"."RUP_BRACKET" := ?; "TECH"."TECH_ASMV"."NEUEREINTRAG" ("P_REC_ASMV" => "r1", "P_RUP" => "r2", "P_RC" => ?); ? := "r1"."ASMV_ID"; ? := "r1"."ASMV_SENDER"; ? := "r1"."ASMV_EMPFAENGER"; ? := "r1"."ASMV_AUSFUEHRUNGSZEITPUNKT"; ? := "r1"."ASMV_CC"; ? := "r1"."ASMV_BCC"; ? := "r1"."ASMV_HEADER"; ? := "r1"."ASMV_TEXT"; ? := "r1"."ASMV_ERSTELLT_VON"; ? := "r1"."ASMV_ERSTELLUNGSZEITPUNKT"; ? := "r1"."ASMV_MIME_TYPE"; ? := "r1"."ASMV_PFAD"; ? := "r1"."ASMV_DATEINAME"; ? := "r1"."ASMV_BLOB"; ? := "r1"."ASMV_EXT_ANBIETER"; ? := "r1"."ASMV_STATUS"; ? := "r1"."ASMV_SMT_ID"; ? := "r2"."RUP_ID"; ? := "r2"."RUP_EREIGNISTYP"; ? := "r2"."RUP_EREIGNISCODE"; ? := "r2"."RUP_EREIGNISTEXT"; ? := "r2"."RUP_ERSTELLT_VON"; ? := "r2"."RUP_ERSTELLUNGSZEITPUNKT"; ? := "r2"."RUP_USERNAME"; ? := "r2"."RUP_APPLIKATION"; ? := "r2"."RUP_PROGRAMMNAME"; ? := "r2"."RUP_PROGRAMMLAUF_ID"; ? := "r2"."RUP_PROZEDURNAME"; ? := "r2"."RUP_CODEBLOCK"; ? := "r2"."RUP_ENTITAET"; ? := "r2"."RUP_PARAMETER"; ? := "r2"."RUP_PROZESS"; ? := "r2"."RUP_TEILPROZESS"; ? := "r2"."RUP_SCHLUESSELTYP"; ? := "r2"."RUP_SCHLUESSEL"; ? := "r2"."RUP_PROGRAMMSTACK"; ? := "r2"."RUP_BRACKET"; end;]; ORA-06550: Zeile 1, Spalte 2373:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'NEUEREINTRAG'
ORA-06550: Zeile 1, Spalte 2373:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'NEUEREINTRAG'
ORA-06550: Zeile 1, Spalte 2373:
PL/SQL: Statement ignored
```

### Steps to reproduce the problem
Call a PLSQL Procedure with a %Rowtype with DATE fields in combination with dateAsTimestamp=true

### Versions

- jOOQ: 3.14.11
- Java: 11
- Database (include vendor): Oracle 18c
- OS: All
- JDBC Driver (include name if inofficial driver): ojdbc8

**Comment by mnisius** [comment]
Hello again, so I did find this proposed solution in the docs: https://www.jooq.org/doc/3.15/manual/reference/reference-data-types/data-types-oracle-date/
I tried it but sadly it did not work for service calls. 

Also I'm a little confused because the proposed code snipped uses deprecated methods. Is this maybe not up to date? 
(the keyword function is deprecated ) 

```java
@Override
public final void sql(BindingSQLContext<Timestamp> ctx) throws SQLException {
        render.keyword("cast").sql('(')
              .visit(val(ctx.value()))
              .sql(' ').keyword("as date").sql(')');
}
```

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll look into this tomorrow morning. Clearly, this should work out of the box with the `dateAsTimestamp=true` flag

> Also I'm a little confused because the proposed code snipped uses deprecated methods. Is this maybe not up to date?
> (the keyword function is deprecated )

Yes that's correct. Will fix that too and make a more up to date example. The replacement would be `visit(keyword("cast"))` or just `sql("cast")`. Unless you have strong opinions about such keywords being customisable via the `Settings.renderKeywordCase`, then there's not much of a difference between the two.

**Comment by lukaseder** [comment]
Manual improvement regarding deprecated API usage: https://github.com/jOOQ/jOOQ/issues/12091

**Comment by lukaseder** [comment]
This is a duplicate of https://github.com/jOOQ/jOOQ/issues/7890

**Comment by lukaseder** [comment]
It works with package content like this:

```sql
  TYPE pls_date_as_timestamp IS RECORD (
    ts TIMESTAMP,
    d DATE
  );
  
  PROCEDURE pls_p_date_as_timestamp (i IN pls_date_as_timestamp, o OUT pls_date_as_timestamp);
  FUNCTION pls_f_date_as_timestamp (i IN pls_date_as_timestamp) RETURN pls_date_as_timestamp;
```

Trying again with `%ROWTYPE`

**Comment by lukaseder** [comment]
This helps reproduce it:

```sql
CREATE TABLE t_12019 (
  ts TIMESTAMP,
  d DATE
)
```

And then:

```sql
  PROCEDURE pls_p_date_as_timestamp_rt (i IN t_12019%rowtype, o OUT t_12019%rowtype);
  FUNCTION pls_f_date_as_timestamp_rt (i IN t_12019%rowtype) RETURN t_12019%rowtype;
```

The test:

```java
    @Test
    public void testOracleDateAsTimestampWithRowtypes() {

        // Normal code generation
        T_12019Record r1 = new T_12019Record();
        assertEquals(r1, PlsObjects.plsPDateAsTimestampRt(configuration(), r1));

        T_12019Record r2 = new T_12019Record(Timestamp.valueOf("2000-01-01 15:30:45.6"), Date.valueOf("2000-01-02"));
        assertEquals(r2, PlsObjects.plsPDateAsTimestampRt(configuration(), r2));

        // With dateAsTimestamp=true
        DateAsTimestampT_12019Record r3 = new DateAsTimestampT_12019Record();
        assertEquals(r3, DateAsTimestampPlsObjects.plsPDateAsTimestampRt(configuration(), r3));

        DateAsTimestampT_12019Record r4In = new DateAsTimestampT_12019Record(Timestamp.valueOf("2000-01-01 15:30:45.6"), Timestamp.valueOf("2000-01-02 15:30:45.6"));
        DateAsTimestampT_12019Record r4Out = new DateAsTimestampT_12019Record(Timestamp.valueOf("2000-01-01 15:30:45.6"), Timestamp.valueOf("2000-01-02 15:30:45.0"));
        assertEquals(r4Out, DateAsTimestampPlsObjects.plsPDateAsTimestampRt(configuration(), r4In));
    }
```

Exception:

```
org.jooq.exception.DataAccessException: SQL [declare type rowtype1 is record("TS" timestamp(6), "D" timestamp); type rowtype2 is record("TS" timestamp(6), "D" timestamp); "r1" rowtype1; "r2" rowtype2; begin "r1"."TS" := ?; "r1"."D" := ?; "TEST"."PLS_OBJECTS"."PLS_P_DATE_AS_TIMESTAMP_RT" ("I" => "r1", "O" => "r2"); ? := "r2"."TS"; ? := "r2"."D"; end;]; ORA-06550: Zeile 1, Spalte 199:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'PLS_P_DATE_AS_TIMESTAMP_RT'
ORA-06550: Zeile 1, Spalte 199:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'PLS_P_DATE_AS_TIMESTAMP_RT'
ORA-06550: Zeile 1, Spalte 199:
PL/SQL: Statement ignored

	at org.jooq_3.15.0-SNAPSHOT.ORACLE18C.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2932)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:643)
	at org.jooq.impl.AbstractRoutine.executeCallableStatement(AbstractRoutine.java:548)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:360)
	at org.jooq.impl.Tools.attach(Tools.java:1228)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:341)
	at org.jooq.test.oracle2.generatedclasses.packages.DateAsTimestampPlsObjects.plsPDateAsTimestampRt(DateAsTimestampPlsObjects.java:790)
	at org.jooq.test.OracleTest.testOracleDateAsTimestampWithRowtypes(OracleTest.java:4672)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.sql.SQLException: ORA-06550: Zeile 1, Spalte 199:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'PLS_P_DATE_AS_TIMESTAMP_RT'
ORA-06550: Zeile 1, Spalte 199:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'PLS_P_DATE_AS_TIMESTAMP_RT'
ORA-06550: Zeile 1, Spalte 199:
PL/SQL: Statement ignored

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:628)
	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:562)
	at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1207)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:727)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:291)
	at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:498)
	at oracle.jdbc.driver.T4CCallableStatement.doOall8(T4CCallableStatement.java:151)
	at oracle.jdbc.driver.T4CCallableStatement.executeForRows(T4CCallableStatement.java:1192)
	at oracle.jdbc.driver.OracleStatement.executeSQLStatement(OracleStatement.java:1747)
	at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1432)
	at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3743)
	at oracle.jdbc.driver.OraclePreparedStatement.execute(OraclePreparedStatement.java:4080)
	at oracle.jdbc.driver.OracleCallableStatement.execute(OracleCallableStatement.java:4279)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.execute(OraclePreparedStatementWrapper.java:1014)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:4018)
	at org.jooq.impl.AbstractRoutine.execute0(AbstractRoutine.java:559)
	at org.jooq.impl.AbstractRoutine.executeCallableStatement(AbstractRoutine.java:519)
	... 36 more
Caused by: Error : 6550, Position : 198, Sql = declare type rowtype1 is record("TS" timestamp(6), "D" timestamp); type rowtype2 is record("TS" timestamp(6), "D" timestamp); "r1" rowtype1; "r2" rowtype2; begin "r1"."TS" := :1 ; "r1"."D" := :2 ; "TEST"."PLS_OBJECTS"."PLS_P_DATE_AS_TIMESTAMP_RT" ("I" => "r1", "O" => "r2"); :3  := "r2"."TS"; :4  := "r2"."D"; end;, OriginalSql = declare type rowtype1 is record("TS" timestamp(6), "D" timestamp); type rowtype2 is record("TS" timestamp(6), "D" timestamp); "r1" rowtype1; "r2" rowtype2; begin "r1"."TS" := ?; "r1"."D" := ?; "TEST"."PLS_OBJECTS"."PLS_P_DATE_AS_TIMESTAMP_RT" ("I" => "r1", "O" => "r2"); ? := "r2"."TS"; ? := "r2"."D"; end;, Error Msg = ORA-06550: Zeile 1, Spalte 199:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'PLS_P_DATE_AS_TIMESTAMP_RT'
ORA-06550: Zeile 1, Spalte 199:
PLS-00306: Falsche Anzahl oder Typen von Argumenten in Aufruf von 'PLS_P_DATE_AS_TIMESTAMP_RT'
ORA-06550: Zeile 1, Spalte 199:
PL/SQL: Statement ignored

	at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:632)
	... 54 more
```

**Comment by lukaseder** [comment]
The fix seems very easy. Just check if a `DataType` has the `org.jooq.impl.DateAsTimestampBinding` and then generate `DATE` as the type declaration, instead of `TIMESTAMP`.

This also affects ordinary DDL type declaration usage, e.g. when we try to generate a `CREATE TABLE` statement from the generated code:

```java
println(create().meta(T_12019).ddl(new DDLExportConfiguration().flags(DDLFlag.TABLE)));
```

This currently produces:

```sql
create table "TEST"."T_12019" (
  "TS" timestamp(6),
  "D" timestamp
);
```

I've created a separate issue for this: https://github.com/jOOQ/jOOQ/issues/12094, and will fix both right away.

**Comment by lukaseder** [comment]
Other affected API:

- `CREATE FUNCTION .. RETURN` type declaration (jOOQ 3.15 only)
- `JSON_TABLE` column declaration (when using field references from generated tables)
- `XMLTABLE` column declaration (when using field references from generated tables)
- PL/SQL variable declaration (e.g. when using the jOOQ equivalent of a `TABLE.COLUMN%TYPE` reference)

In all of the above, a workaround is easy: Provide a `SQLDataType.DATE` reference explicitly

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.12 (#12095). Thanks again for your report!

**Comment by lukaseder** [comment]
3.14.12 has been released including this fix
